lib/core/utils/get-selector.js

```diff
@@ -47,174 +185,162 @@ function getNthChildString (elm, selector) {
+        });
+      }
+    });
+  }
+  return retVal.sort(function (a, b) {
+    return (a.count < b.count) ? -1 : (a.count === b.count) ? 0 : 1;
+  });
+}
+
+/**
+ * Get an array of features (as CSS selectors) that are likely to uniquely
+ * describe an element
+ *
+ */
+function getDistinctAttributes (elm) {
+  let atts = uncommonAttributes(elm, axe._selectorData.attributes);
+
+  if (atts.length) {
+    // If we have a unique attribute, use that
+    if (atts[0].count === 1) {
+      return [atts[0]];
     }
+    return atts.slice(0, 3);
+  } else {
+    return [];
+  }
+}
 
-    const feature = func(elm, props);
-    if (feature) {
-      if (!feature[0].match(/[a-z]/)) {
-        features.push(feature);
-      } else {
-        features.unshift(feature);
-      }
+function getThreeLeastCommonFeatures(elm) {
+  let features = getDistinctClassList(elm).concat(getDistinctAttributes(elm));
+  let selector = '';
+  
+  // sort by least common
+  features.sort((a, b) => {
+    return (a.count < b.count) ? -1 : (a.count === b.count) ? 0 : 1;
+  });
+
+  // select three least common
+  features = features.slice(0, 3);
+
+  // if no class, add the tag
+  if (!features.some((feat) => { return feat.species === 'class'; })) {
+    // has no class
+    selector = getBaseSelector(elm);
+  }
+
+  // put the classes at the front of the selector
+  features.sort((a, b) => {
+    return (a.species !== b.species && a.species === 'class') ? -1 : (a.species === b.species) ? 0 : 1;
+  });
+
+  // construct the return value
+  return selector += features.reduce((val, feat) => {
+    switch(feat.species) {
+      case 'class':
+        return val + '.' + feat.name;
+      case 'attribute':
+        return val + '[' + feat.name + ']';
     }
-    return features;
-  }, []);
+    return val; // should never happen
+  }, '');
 }
 
 function generateSelector (elm, options, doc) {
   //jshint maxstatements: 19
-  let selector, addParent;
-  let { isUnique = false } = options;
-  const idSelector = createSelector.getElmId(elm);
+  //jshint loopfunc:true
   const {
-    featureCount = 2,
-    minDepth = 1,
-    toRoot = false,
-    childSelectors = []
+    toRoot = false
   } = options;
+  let selector;
+  let similar;
 
-  if (idSelector) {
-    selector = idSelector;
-    isUnique = true;
-
-  } else {
-    selector = getElmFeatures(elm, featureCount).join('');
-    selector += getNthChildString(elm, selector);
-    isUnique = options.isUnique || doc.querySelectorAll(selector).length === 1;
-
-    // For the odd case that document doesn't have a unique selector
-    if (!isUnique && elm === document.documentElement) {
-      // todo: figure out what to do for shadow DOM
-      selector += ':root';
+  /**
+   * Try to find a unique selector by filtering out all the clashing
+   * nodes by adding ancestor selectors iteratively.
+   * This loop is much faster than recursing and using querySelectorAll
+   */
+  do {
+    let features = getElmId(elm);
```

Comment:
does not make sense because the ids are only unique within a shadow boundary and that would add a lot of extra processing. I think its better to just let querySelectorAll figure that out for us.